cache/clock_cache.h

```diff
@@ -565,53 +567,86 @@ class ClockHandleTable {
-  // decrements all displacements, starting from the 0-th probe.
+  // Extracts the element information from a handle (src), and assigns it
+  // to a hash table slot (dst). Doesn't touch displacements and refs,
+  // which are maintained by the hash table algorithm.
+  void Assign(ClockHandle* dst, ClockHandle* src);
+
+  // Returns the first slot in the probe sequence, starting from the given
+  // probe number, with a handle e such that match(e) is true. At every
+  // step, the function first tests whether match(e) holds. If this is false,
+  // it evaluates abort(e) to decide whether the search should be aborted,
+  // and in the affirmative returns -1. For every handle e probed except
+  // the last one, the function runs update(e).
+  // The probe parameter is modified as follows. We say a probe to a handle
+  // e is aborting if match(e) is false and abort(e) is true. Then the final
+  // value of probe is one more than the last non-aborting probe during the
+  // call. This is so that that the variable can be used to keep track of
+  // progress across consecutive calls to FindSlot.
+  inline ClockHandle* FindSlot(const Slice& key,
+                               std::function<bool(ClockHandle*)> match,
+                               std::function<bool(ClockHandle*)> stop,
+                               std::function<void(ClockHandle*)> update,
+                               uint32_t& probe);
+
+  // Returns an available slot for the given key. All copies of the
+  // key found along the probing sequence until an available slot is
+  // found are marked for deletion. On each of them, a deletion is
+  // attempted, and when the attempt succeeds the slot is assigned to
+  // the new copy of the element.
+  ClockHandle* FindAvailableSlot(const Slice& key, uint32_t hash,
+                                 uint32_t& probe,
+                                 autovector<ClockHandle>* deleted);
+
+  // After a failed FindSlot call (i.e., with answer -1) in
+  // FindAvailableSlot, this function fixes all displacements
+  // starting from the 0-th probe.
   void Rollback(const Slice& key, uint32_t probe);
 
   // Number of hash bits used for table index.
   // The size of the table is 1 << length_bits_.
-  int length_bits_;
+  const int length_bits_;
 
   // For faster computation of ModTableSize.
   const uint32_t length_bits_mask_;
 
-  // Number of elements in the table.
-  uint32_t occupancy_;
-
   // Maximum number of elements the user can store in the table.
-  uint32_t occupancy_limit_;
+  const uint32_t occupancy_limit_;
 
+  // Maximum total charge of all elements stored in the table.
+  const size_t capacity_;
+
+  // ------------^^^^^^^^^^^^^-----------
+  // Not frequently modified data members
+  // ------------------------------------
+  //
+  // We separate data members that are updated frequently from the ones that
+  // are not frequently updated so that they don't share the same cache line
```

Comment:
How is cacheline separation being ensured? Also, shouldn't each frequently modified data member be on a separate cacheline? And array_ is not frequently modified I believe?

Reply:
To be honest, I ported this from LRUCache, and didn't give it much thought, so thank you for asking this question. I think we want to avoid reads on some field X to trigger the cache coherence mechanism only because it shares the cache line with other modified field Y, not because the operation reading X actually needs to use Y. In our case, Lookup uses array_; Release and Erase use array_, occupancy_ and usage_; and Insert uses array_, occupancy_, usage_ and clock_pointer_. So I think we want 3 different cache lines: array_; occupancy_ and usage_; and clock_pointer_.

As to how is cache line separation ensured: I think we should be aligning the fields.